Bdd participant setup

MattMachin opened the pull request.

MattMachin committed:
BDD for participant setup, send results and view participants

MattMachin committed:
Update viewParticipants.feature

Xiaoxue-Zheng commented on mobile/tests/e2e/specs/participantSetup.feature:
```diff
@@ -0,0 +1,20 @@
+
+    Scenario: Reminder of unique code
+        Given: The user has already opened the app and generated a code
+        And: The user wants a reminder of the unique code
```
"The user wants" is not a good description

MattMachin replied:
I have changed this.

Xiaoxue-Zheng commented:
When BDD integrate with end 2 end test, each sentence in the BDD will correspond to the operations or element description on the browser, Cypress will support interacting with browsers. 

To better translate BDD to Cypress lauguage(interacting with browsers), I would like to suggest some rules when writing BDD.

Given:
Should contain verbs like: open app/visit url/sign up/sign in/view
One verb in each sentense is prefered.

When:
Should contain verbs like: open/click/select/enter/choose/go back/tap/start
One verb in each sentense is prefered.

Then:
Should contain descriptions like: ticked/opened/visible/displayed(shown)/returned/notified/xxx is correct
Only one element is discribed in one sentense is prefered.
In some cases, these description words may appear in "When" and "Given" sentence, to make sure the steps make sense

MattMachin committed:
Updated after review

All BDD updated in response to review comments. I'm a bit unsure about the "sendResults.feature" as some of the then clauses relate to storing data rather than something the user sees. Perhaps this could be written differently? Suggestions would be welcomed.

MattMachin commented on mobile/tests/e2e/specs/participantSetup.feature:
```diff
@@ -0,0 +1,20 @@
+Story: Create unique code in app 
```
I have updated this

MattMachin commented on mobile/tests/e2e/specs/participantSetup.feature:
```diff
@@ -0,0 +1,20 @@
+        When: They open the app
+        Then: A unique code will be generated that is stored in the app
+        And: The user will be shown the code on the screen
+        And: The code will remain on screen until the user dismisses it
```
I have updated this (all occurrences)

MattMachin commented on mobile/tests/e2e/specs/participantSetup.feature:
```diff
@@ -0,0 +1,20 @@
+        Given: The user has already opened the app and generated a code
+        When: They open the app
+        Then: No unique code will be generated
+        And: The user can interact with the normal app features
```
I have changed this but maybe we need to agree what should be on the homescreen?

MattMachin commented on mobile/tests/e2e/specs/sendResults.feature:
```diff
@@ -0,0 +1,9 @@
+Feature: Send Results
+
+Story: Send results from app to server
```
I have updated this

MattMachin commented on mobile/tests/e2e/specs/participantSetup.feature:
```diff
@@ -0,0 +1,20 @@
+    Scenario: Reminder of unique code
+        Given: The user has already opened the app and generated a code
+        And: The user wants a reminder of the unique code
+        When: They select the "unique code" menu item
```
We should probably discuss this.

MattMachin committed:
Update following second review

MattMachin commented on mobile/tests/e2e/specs/sendResults.feature:
```diff
@@ -0,0 +1,10 @@
+    Then A message will be displayed to the user to thank them for completing the questions
+    And The app shall send the data to the server, along with the unique code
+    And The server will store received data for the correct participant
+    And The unique code will be used to identify the correct participant
```
Done

MattMachin commented on web/tests/e2e/specs/viewParticipants.feature:
```diff
@@ -0,0 +1,14 @@
+    When The researcher logs in to the web interface
+    Then They will see a "summary view" of existing participants
+    And The view shall be ordered by participant id.
+    And For each participant the view shall include participant id, unique code and a link (e.g. button) to view data
```
Done
